File: app/src/main/java/zmuzik/slidingpuzzle2/common/ShakeDetector.java
package zmuzik.slidingpuzzle2.common;

import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;

public class ShakeDetector implements SensorEventListener {

    private static final float SHAKE_FORCE_THRESHOLD = 2f;
    private static final int SHAKE_SLOP_TIME_MS = 500;

    private final Context mContext;
    private final SensorManager mSensorManager;
    private final Sensor mAccelerometer;
    private OnShakeListener mListener;
    private long mShakeTimestamp;
    private boolean mIsRegistered;

    public ShakeDetector(Context context) {
        mContext = context;
        mSensorManager = (SensorManager) mContext.getSystemService(Context.SENSOR_SERVICE);
        mAccelerometer = mSensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);
    }

    public void setOnShakeListener(OnShakeListener listener) {
        this.mListener = listener;
    }

    public void register() {
        if (!mIsRegistered) {
            mSensorManager.registerListener(this, mAccelerometer, SensorManager.SENSOR_DELAY_UI);
            mIsRegistered = true;
        }
    }

    public void unRegister() {
        if (mIsRegistered) {
            mSensorManager.unregisterListener(this);
            mIsRegistered = false;
        }
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
    }

    @Override
    public void onSensorChanged(SensorEvent event) {
        if (mListener != null) {
            float x = event.values[0];
            float y = event.values[1];
            float z = event.values[2];

            float gX = x / SensorManager.GRAVITY_EARTH;
            float gY = y / SensorManager.GRAVITY_EARTH;
            float gZ = z / SensorManager.GRAVITY_EARTH;

            double totalForce = Math.sqrt(gX * gX + gY * gY + gZ * gZ);

            if (totalForce > SHAKE_FORCE_THRESHOLD) {
                final long now = System.currentTimeMillis();
                if (mShakeTimestamp + SHAKE_SLOP_TIME_MS > now) return;
                mShakeTimestamp = now;
                mListener.onShake();
            }
        }
    }

    public interface OnShakeListener {
        public void onShake();
    }
}

File: app/src/main/java/zmuzik/slidingpuzzle2/common/ShakeDetector.kt
package zmuzik.slidingpuzzle2.common

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import java.lang.ref.WeakReference
import kotlin.math.sqrt

class ShakeDetector(private val context: Context) : SensorEventListener {

    private val shakeForceThreshold = 2f
    private val shakeSlopTimeMs = 500

    private val sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
    private val accelerometer = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
    private var listener: WeakReference<OnShakeListener>? = null
    private var shakeTimestamp: Long = 0
    private var isRegistered: Boolean = false

    fun setOnShakeListener(listener: OnShakeListener) {
        this.listener = WeakReference(listener)
    }

    fun register() {
        if (!isRegistered) {
            sensorManager.registerListener(this, accelerometer, SensorManager.SENSOR_DELAY_UI)
            isRegistered = true
        }
    }

    fun unRegister() {
        if (isRegistered) {
            sensorManager.unregisterListener(this)
            isRegistered = false
        }
    }

    override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}

    override fun onSensorChanged(event: SensorEvent) {
        listener?.get()?.let {
            val x = event.values[0]
            val y = event.values[1]
            val z = event.values[2]

            val gX = x / SensorManager.GRAVITY_EARTH
            val gY = y / SensorManager.GRAVITY_EARTH
            val gZ = z / SensorManager.GRAVITY_EARTH

            val totalForce = sqrt((gX * gX + gY * gY + gZ * gZ).toDouble())

            if (totalForce > shakeForceThreshold) {
                val now = System.currentTimeMillis()
                if (shakeTimestamp + shakeSlopTimeMs > now) return
                shakeTimestamp = now
                it.onShake()
            }
        }
    }

    interface OnShakeListener {
        fun onShake()
    }
}
